scheduling-utils/src/handshake/server.rs

```diff
@@ -0,0 +1,296 @@
+    std::{
+        ffi::CStr,
+        fs::File,
+        io::{IoSlice, Read, Write},
+        os::{
+            fd::{AsRawFd, FromRawFd},
+            unix::net::{UnixListener, UnixStream},
+        },
+        path::Path,
+        time::{Duration, Instant},
+    },
+    thiserror::Error,
+};
+
+type ShaqError = shaq::error::Error;
+type RtsAllocError = rts_alloc::error::Error;
+
+const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(1);
+const SHMEM_NAME: &CStr = c"/agave-scheduler-bindings";
+
+pub struct Server {
+    listener: UnixListener,
+
+    buffer: [u8; 1024],
+    buffer_len: usize,
+}
+
+impl Server {
+    pub fn new(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
+        let listener = UnixListener::bind(path)?;
+
+        Ok(Self {
+            listener,
+            buffer: [0; 1024],
+            buffer_len: 0,
+        })
+    }
+
+    pub fn accept(&mut self) -> Result<Session, HandshakeError> {
+        // Clear the buffer.
+        self.buffer_len = 0;
+
+        // Wait for next stream.
+        let (mut stream, _) = self.listener.accept()?;
+        stream.set_read_timeout(Some(HANDSHAKE_TIMEOUT))?;
+
+        match self.handle_logon(&mut stream) {
+            Ok(session) => Ok(session),
+            Err(err) => {
+                let reason = err.to_string();
+                let reason_len = u8::try_from(reason.len()).unwrap_or(u8::MAX);
+
+                let buffer_len = 2usize.wrapping_add(usize::from(reason_len));
+                self.buffer[0] = LOGON_FAILURE;
+                self.buffer[1] = reason_len;
+                self.buffer[2..buffer_len]
+                    .copy_from_slice(&reason.as_bytes()[..usize::from(reason_len)]);
+
+                stream.set_nonblocking(true)?;
+                // NB: Caller will still error out even if our write fails so it's fine to ignore the
+                // result.
+                let _ = stream.write(&self.buffer[..buffer_len])?;
+
+                Err(err)
+            }
+        }
+    }
+
+    fn handle_logon(&mut self, stream: &mut UnixStream) -> Result<Session, HandshakeError> {
+        // Read the logon message.
+        let handshake_start = Instant::now();
+        while self.buffer_len < self.buffer.len() {
+            let read = stream.read(&mut self.buffer[self.buffer_len..])?;
+            if read == 0 {
+                return Err(HandshakeError::EofDuringHandshake);
+            }
+
+            // SAFETY: We cannot read a value greater than buffer.len() which itself is a usize.
+            self.buffer_len = self.buffer_len.wrapping_add(read);
+
+            if handshake_start.elapsed() > HANDSHAKE_TIMEOUT {
+                return Err(HandshakeError::Timeout);
+            }
+        }
+
+        // Ensure exact version match, version will be bumped any time a backwards incompatible
+        // change is made to handshake/shared memory objects.
+        let version = u64::from_le_bytes(*array_ref![self.buffer, 0, 8]);
```

Comment:
sorry to be a stickler, but I hate adding dependencies for what is effectively just a ptr read.

```rust
    use std::convert::TryInto;
    u64::from_le_bytes(buffer[0..8].try_into().unwrap())
```

O3 this is literally just a mov, array_ref save us no work here but increases compilation time for everyone who uses the utils library.

Reply:
Makes sense